packages/daemon/src/host.js

```diff
@@ -132,48 +132,48 @@ export const makeHostMaker = ({
      * @returns {Promise<{formulaIdentifier: string, value: Promise<import('./types.js').EndoGuest>}>}
      */
     const makeGuest = async (petName, { introducedNames = {} } = {}) => {
-      /** @type {string | undefined} */
-      let formulaIdentifier;
       if (petName !== undefined) {
-        formulaIdentifier = petStore.identifyLocal(petName);
-      }
+        const formulaIdentifier = petStore.identifyLocal(petName);
+        if (formulaIdentifier !== undefined) {
+          if (formulaIdentifier.startsWith('guest:')) {
+            throw new Error(
+              `Existing pet name does not designate a guest powers capability: ${q(
+                petName,
+              )}`,
+            );
+          }
 
-      if (formulaIdentifier === undefined) {
-        const { formulaIdentifier: hostHandleFormulaIdentifier } =
-          await makeNewHandleForSelf();
-        const { value, formulaIdentifier: guestFormulaIdentifier } =
-          // Behold, recursion:
-          // eslint-disable-next-line no-use-before-define
-          await incarnateGuest(hostHandleFormulaIdentifier);
-        if (petName !== undefined) {
-          assertPetName(petName);
-          await petStore.write(petName, guestFormulaIdentifier);
+          // TODO: Should this be awaited?
+          introduceNamesToParty(formulaIdentifier, introducedNames);
+          const guestController =
+            provideControllerForFormulaIdentifier(formulaIdentifier);
+          return {
+            formulaIdentifier,
+            value: /** @type {Promise<import('./types.js').EndoGuest>} */ (
+              guestController.external
+            ),
+          };
         }
-
-        return {
-          value: Promise.resolve(value),
-          formulaIdentifier: guestFormulaIdentifier,
-        };
-      } else if (!formulaIdentifier.startsWith('guest:')) {
-        throw new Error(
-          `Existing pet name does not designate a guest powers capability: ${q(
-            petName,
-          )}`,
-        );
       }
 
-      const newGuestController =
-        /** @type {import('./types.js').Controller<>} */ (
-          provideControllerForFormulaIdentifier(formulaIdentifier)
+      /** @type {AsyncHooks<import('./types.js').GuestHookParams>} */
+      const hooks = makeAsyncHooks();
+      if (petName !== undefined) {
+        hooks.add(identifiers =>
+          petStore.write(petName, identifiers.guestFormulaIdentifier),
         );
-      if (introducedNames !== undefined) {
-        introduceNamesToNewHostOrGuest(newGuestController, introducedNames);
       }
+
+      const { value, formulaIdentifier } =
+        // Behold, recursion:
+        // eslint-disable-next-line no-use-before-define
+        await incarnateGuest(hostFormulaIdentifier, hooks);
+      // TODO: Should this be awaited?
+      introduceNamesToParty(formulaIdentifier, introducedNames);
+
       return {
+        value: Promise.resolve(value),
```

Comment:
Is the type system demanding we wrap these? Maybe we should use `Awaited<T>` somewhere.

Reply:
yes

Reply:
The type system does demand this. Since this was already here, I'll leave it for now. Should we create an issue?

Reply:
Meh. Not big enough to track.